src/sample-executors/SwapRouter02ExecutorWithPermit.sol

```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.8.0;
+
+import {Owned} from "solmate/src/auth/Owned.sol";
+import {SafeTransferLib} from "solmate/src/utils/SafeTransferLib.sol";
+import {ERC20} from "solmate/src/tokens/ERC20.sol";
+import {WETH} from "solmate/src/tokens/WETH.sol";
+import {IReactorCallback} from "../interfaces/IReactorCallback.sol";
+import {IReactor} from "../interfaces/IReactor.sol";
+import {CurrencyLibrary} from "../lib/CurrencyLibrary.sol";
+import {ResolvedOrder, OutputToken, SignedOrder} from "../base/ReactorStructs.sol";
+import {ISwapRouter02} from "../external/ISwapRouter02.sol";
+import {SwapRouter02Executor} from "./SwapRouter02Executor.sol";
+
+/// @notice A fill contract that uses SwapRouter02 to execute trades
+contract SwapRouter02ExecutorWithPermit is SwapRouter02Executor {
+    constructor(address _whitelistedCaller, IReactor _reactor, address _owner, ISwapRouter02 _swapRouter02)
+        SwapRouter02Executor(_whitelistedCaller, _reactor, _owner, _swapRouter02)
+    {}
+
+    /// @notice assume that we already have all output tokens
+    /// @notice the reactor performs no verification that the user's signed permit is executed correctly
+    ///         e.g. if the necessary approvals are already set, a filler can call this function or the standard execute function to fill the order
+    /// @dev assume 2612 permit is collected offchain
+    function executeWithPermit(SignedOrder calldata order, bytes calldata callbackData, bytes calldata permitData)
+        external
+        onlyWhitelistedCaller
+    {
+        _permit(permitData);
+        reactor.executeWithCallback(order, callbackData);
+    }
+
+    /// @notice assume that we already have all output tokens
+    /// @dev assume 2612 permits are collected offchain
+    function executeBatchWithPermit(
+        SignedOrder[] calldata orders,
+        bytes calldata callbackData,
+        bytes[] calldata permitData
```

Comment:
doing array of permitData for executeBatch and single for execute single doesn't seem like a perfect matchup - like in the batch of orders they might not all need permits, so just cuz its a list of orders doesnt mean its necessarily a list of permits.. likewise maybe for executesingle we might want to send multiple permits for some reason?

i think we should just decouple it and have a multicall and permit function so the permits can be layered on granularly

Reply:
maybe we should make a base contract that other executors can inherit from that has this shared logic? wdyt

Reply:
ooo good point about the multiple orders not all needing permits, it would be weird to have the inputs be different lengths...

how would you layer permits on granularly? are you thinking of bundling the orders together with any applicable permits like [ (order, permits[]), (order, permits[]), ...] ?